metrics/tags.go

```diff
@@ -3,12 +3,162 @@ package metrics
 import (
 	"bytes"
 	"encoding/json"
+	"errors"
 	"sort"
 	"strings"
+
+	"github.com/mailru/easyjson"
+	"github.com/mailru/easyjson/jlexer"
+	"github.com/mailru/easyjson/jwriter"
+	"github.com/mstoykov/atlas"
 )
 
+// A TagSet represents an immutable set of metric tags. For the efficient and
+// thread-safe storage of the key=value tag pairs, it uses the
+// https://github.com/mstoykov/atlas data structure.
+//
+// Assuming all tag sets start from the same root (see Registry.RootTagSet()),
+// you can compare *TagSet values of different metric Samples with the `==` to
+// check if they have the same tags, and you can also use *TagSet values for map
+// indexes and caching.
+//
+// See also the TimeSeries type for comparing a Sample's {metric+tags} for
+// equality at the same time.
+type TagSet atlas.Node
+
+// With returns another TagSet object that contains the combination of the
+// current receiver tags and the name=value tag from its parameters.
+//
+// It doesn't modify the receiver, it will either return an already existing
+// TagSet with these tags, if it exists, or create a new TagSet with them and
+// return it.
+//
+// If a tag with the specified name already exists in the set, it will be
+// overwritten with the new value in the returned set.
+func (ts *TagSet) With(name, value string) *TagSet {
+	return (*TagSet)(((*atlas.Node)(ts)).AddLink(name, value))
+}
+
+// Without returns another TagSet object that contains all of the tags from the
+// existing TagSet except the one with the given key.
+//
+// It doesn't modify the receiver, it will either return an already existing
+// TagSet with these tags, if it exists, or create a new TagSet with them and
+// return it.
+//
+// If a tag with the specified name doesn't exist in the set, it will return the
+// receiver.
+func (ts *TagSet) Without(name string) *TagSet {
```

Comment:
Same nitpick as for `With` 🐻

Reply:
Resolving for the same reason as https://github.com/grafana/k6/pull/2594#discussion_r946880613